File: Labs/Lab 6/Lab6.java
import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;

public class Lab6 {
    public boolean canFinish(int numCourses, int[][] prerequisites) {
        int[] adjacencyCount = new int[numCourses];
        for(int[] prerequisitePair : prerequisites){
            int needsPrerequisite = prerequisitePair[0];
            adjacencyCount[needsPrerequisite]++;
        }
        Queue<Integer> queue = new LinkedList<>();
        for(int i = 0; i < adjacencyCount.length; i++){
            if (adjacencyCount[i] == 0){
                queue.add(i);
            }
        }
        List<Integer> visited = new ArrayList<>();
        while(!queue.isEmpty()){
            int temp = queue.poll();
            for (int[] prerequisitePair: prerequisites){
                if (prerequisitePair[1] == temp){
                    adjacencyCount[prerequisitePair[0]]--;
                    if (adjacencyCount[prerequisitePair[0]] == 0){
                        queue.add(prerequisitePair[0]);
                    }
                }
            }
            visited.add(temp);
        }
        if (visited.size() == numCourses){
            return true;
        }
        else{
            return false;
        }
    }
}

File: Labs/Lab 6/Lab6.py
from collections import deque
class Lab6(object):
    def canFinish(numCourses, prerequisites):
        adjacencyCount = [0] * numCourses
        for prerequisitePair in prerequisites:
            needsPrerequisite = prerequisitePair[0]
            adjacencyCount[needsPrerequisite] += 1
        queue = deque()
        for i in range(len(adjacencyCount)):
            if adjacencyCount[i] == 0:
                queue.append(i)
        visited = []
        while queue:
            temp = queue.pop()
            for prerequisitePair in prerequisites:
                if prerequisitePair[1] == temp:
                    adjacencyCount[prerequisitePair[0]] -= 1
                    if adjacencyCount[prerequisitePair[0]] == 0:
                        queue.append(prerequisitePair[0])
            visited.append(temp)
        return len(visited) == numCourses